ecosystem/python/sdk/aptos_sdk/account_sequence_number.py

```diff
@@ -0,0 +1,178 @@
+    * This is co-routine safe, that is many async tasks can be reading from this concurrently.
+    * The synchronize method will create a barrier that prevents additional next_sequence_number
+      calls until it is complete.
+    * This only manages the distribution of sequence numbers it does not help handle transaction
+      failures.
+    """
+
+    client: RestClient
+    account: AccountAddress
+    lock = asyncio.Lock
+
+    maximum_in_flight: int = 100
+    maximum_wait_time = 30
+    sleep_time = 0.01
+
+    last_committed_number: Optional[int]
+    current_number: Optional[int]
+
+    def __init__(self, client: RestClient, account: AccountAddress):
+        self.client = client
+        self.account = account
+        self.lock = asyncio.Lock()
+
+        self.last_uncommitted_number = None
+        self.current_number = None
+
+    async def next_sequence_number(self, block: bool = True) -> Optional[int]:
+        """
+        Returns the next sequence number available on this account. This leverages a lock to
+        guarantee first-in, first-out ordering of requests.
+        """
+        await self.lock.acquire()
+        try:
+            if self.last_uncommitted_number is None or self.current_number is None:
+                await self.initialize()
+            if (
+                self.current_number - self.last_uncommitted_number
+                >= self.maximum_in_flight
+            ):
+                await self.__update()
+                start_time = time.time()
+                while (
+                    self.current_number - self.last_uncommitted_number
+                    >= self.maximum_in_flight
+                ):
+                    if not block:
+                        return None
+                    await asyncio.sleep(self.sleep_time)
+                    if time.time() - start_time > self.maximum_wait_time:
+                        logging.warn(
+                            f"Waited over 30 seconds for a transaction to commit, resyncing {self.account.address().hex()}"
+                        )
+                        await self.initialize()
+                    else:
+                        await self.__update()
+            next_number = self.current_number
+            self.current_number += 1
+        finally:
+            self.lock.release()
+        return next_number
+
+    async def initialize(self):
+        """Optional initializer. called by next_sequence_number if not called prior."""
+        self.current_number = await self.__current_sequence_number()
+        self.last_uncommitted_number = self.current_number
+
+    async def synchronize(self):
+        """
+        Poll the network until all submitted transactions have either been committed or until
+        the maximum wait time has elapsed. This will prevent any calls to next_sequence_number
+        until this called has returned.
+        """
+        if self.last_uncommitted_number == self.current_number:
+            return
```

Comment:
This is a short cut, but maybe want to provide an input flag to sync regardless (if for some reason you've got something else also handling transactions).